Add dry_run mode to EnvProvider

wata-gh opened the pull request:
I wanted to add dry_run option to EnvProvider constructor.
But, for backward compatibility I decided to add method to pass
dry run flag.
If it is not necessary to care about backward compatibility, I would
like pass dry run flag by constructor.

---

Dry run mode for EnvProvider would help to check secret values would correctly expand before deploying.

wata-gh commented:
I created  pull request for hako-vault to support dry-run mode.
https://github.com/eagletmt/hako-vault/pull/1

eagletmt commented:
Adding dry-run mode to EnvProvider looks good idea 🆒 

But `show-yaml` subcommand doesn't look suitable place for it.
Expanding variables embedded in `env` isn't YAML extension while `show-yaml` is expected to perform YAML extensions (i.e., `!include`).

What do you think of using `deploy --dry-run` ?
Since you modified Container#expand_env which is called in `deploy --dry-run`, `deploy --dry-run` can be used to validate embedded variables.

wata-gh commented:
Validate secret values at `deploy --dry-run` seems good idea. 👍 
I implemented `print_expanded_yaml` and call from deploy method in Commander class.
Also renamed method to validate.

eagletmt commented on lib/hako/container.rb:
```diff
@@ -84,10 +84,10 @@ def log_configuration
       provider_types = env.delete(PROVIDERS_KEY) || []
-      if @dry_run
+      providers = load_providers(provider_types)
+      if @dry_run && providers.any? { |provider| !provider.validatable? }
```
Validating in `EnvProvider#expand` looks weird to me...
What do you think of adding another method like below?

```ruby
expander = EnvExpander.new(providers)
if @dry_run
  expander.validate_undefined_variables!(env)
  env
else
  expander.expand(env)
end
```

```ruby
class EnvExpander
  def validate_undefined_variables!(env)
    if @providers.any? { |provider| !provider.respond_to?(:ask_keys) }
      # Unable to validate
      return true
    end
    variables = ...
    @providers.each do |provider|
      if variables.empty?
        break
      end
      provider.ask_keys(variables.to_a).each do |var|
        variables.delete(var)
      end
    end
    unless variables.empty?
      raise ExpansionError
    end
    true
  end
end
```

wata-gh commented on lib/hako/commander.rb:
```diff
@@ -13,12 +13,30 @@ def initialize(app)
     def deploy(force: false, tag: 'latest', dry_run: false, timeout:)
       containers = load_containers(tag, dry_run: dry_run)
+      if dry_run
+        print_expanded_yaml(containers)
```
I would like to see which EnvProvider expanded the secret value.
Just in case multiple expander has same key.

eagletmt replied:
`print_expanded_yaml` doesn't seem to tell which provider expands the variable...
Could you show me example output?

wata-gh commented on lib/hako/commander.rb:
```diff
@@ -13,12 +13,30 @@ def initialize(app)
       @app = app
     end
 
+    def show_yaml(dry_run: false)
```
Sorry.
Removed.
https://github.com/eagletmt/hako/pull/26/commits/b1bd6c4e0b45298f393a0e344ab910df21a2bf91

wata-gh commented:
Seems printing and validation is not what you wanted.
I will close this PR.
Thanks for your time. ;-)
